### PR TITLE
REF/TYP: define methods non-dynamically for SparseArray

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/sparse/array.py [hunk 1]
```diff
@@ -1444,56 +1431,60 @@ def sparse_arithmetic_method(self, other):
...
+    _logical_method = _cmp_method
 
-            if isinstance(other, SparseArray):
-                return _sparse_array_op(self, other, op, op_name)
-            else:
-                with np.errstate(all="ignore"):
-                    fill_value = op(self.fill_value, other)
-                    result = op(self.sp_values, other)
+    @unpack_zerodim_and_defer("__and__")
+    def __and__(self, other) -> "SparseArray":
```
Comment (simonjayhawkins) [inline pandas/core/arrays/sparse/array.py:1463]: with #36964 merged, can these now be removed?

File: pandas/core/arrays/sparse/array.py [hunk 2]
```diff
@@ -1444,56 +1431,60 @@ def sparse_arithmetic_method(self, other):
...
-                )
+    @unpack_zerodim_and_defer("__or__")
+    def __or__(self, other) -> "SparseArray":
+        return self._logical_method(other, operator.or_)
 
-        name = f"__{op.__name__}__"
-        return compat.set_function_name(cmp_method, name, cls)
+    @unpack_zerodim_and_defer("__xor__")
+    def __xor__(self, other) -> "SparseArray":
+        return self._logical_method(other, operator.xor)
```
Comment (simonjayhawkins) [inline pandas/core/arrays/sparse/array.py:1472]: this was using _create_arithmetic_method b4 whereas `and` and `or` were using _create_comparison_method. Any user facing changes?

File: pandas/core/arrays/sparse/array.py [hunk 3]
```diff
@@ -796,7 +795,6 @@ def __getitem__(self, key):
             key = check_array_indexer(self, key)
 
             if com.is_bool_indexer(key):
-                key = check_bool_indexer(self, key)
```
Comment (simonjayhawkins) [inline pandas/core/arrays/sparse/array.py:799]: not needed?

File: pandas/core/arrays/sparse/array.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/sparse/array.py]: yes, updated

**Pull request merged**